configs/myconfigs/sampling/Sampling.py

```diff
@@ -0,0 +1,361 @@
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# Authors: Brian Coutinho, Keshav Mathur, Jason Lowe-Power
+
+""" Statistical Sampling library for gem5:
+
+    This library provides support for performing statistical sampling
+    on multi-threaded workloads.  See the accompanying example script
+    files for examples on using this library.
+
+    ----------------------
+    What this module expects:
+    The system object must have a three cpu models named as follows-
+
+      system.cpu : should be a list of type KVM cpu to enable KVM
+                   fast-forward mode
+      system.atomicCpu : should be a matching sized list of the
+                   AtomicSimpleCPU type for warming up caches etc.
+      system.timingCPU: should contain instances of the cpu type
+                   you wish to use in the experiments
+
+    Secondly, if you simulate a multi-threaded application the system
+    should have some componenet that adds a small amount of non-determinism
+    to each simulation. In the example scripts we add a small random delay
+    on every last level cache miss to create this effect. The module
+    changes the random seed for each simulation at a sample point in order
+    to cover all divergent paths.
+    ----------------------
+
+    Please look at the technical report below for details on how to
+    determine sampling paramters and a recommended flowchart for the
+    sampling process
+    <TBD>
+"""
+
+import os
+import sys
+import time
+
+import m5
+from m5.objects import *
+from m5.ticks import fromSeconds
+from m5.util.convert import toLatency
+
+sys.path.append('configs/common/') # For the next line...
+import SimpleOpts
+
+
+SimpleOpts.add_option("--warmup_time", default="10ms",
+                      help="Simulated functional warmup time in ms")
+SimpleOpts.add_option("--detailed_warmup_time", default="10us",
+                      help="Simulated detailed warmup time in ms")
+SimpleOpts.add_option("--detailed_time", default="1ms",
+                      help="Simulated time in ms for detailed measurements, \
+                          (size of measurment window)")
+
+from sim_tools import *
+
+
+_valid_system = False
+sample_insts = []            #-- list storing instructions for sample points
+_sample = 0                  #-- current sample index
+
+
+def generateSamples(system, opts, instructions, samples, \
+                    start_insts):
+    """ Generates a fixed number of random sample points that can
+        be used by other methods of this module.
+
+        @param system the system we are running
+        @param opts command line options as an instance of optparse
+        @param instructions total instructions in the ROI (summed over all
+               CPUs)
+        @param samples number of sample points in the ROI
+        @param start_insts instructions executed till the start of roi
```

Comment:
The function does not return anything, the module's internal variable sample_insts gets updated, I can mention this in the comment/doc.